clippy_lints/src/confusing_xor_and_pow.rs

```diff
@@ -0,0 +1,104 @@
+    /// ```
+    #[clippy::version = "1.65.0"]
+    pub CONFUSING_XOR_AND_POW,
+    pedantic,
+    "the `^` operator is being used as exponentiation"
+}
+declare_lint_pass!(ConfusingXorAndPow => [CONFUSING_XOR_AND_POW]);
+
+impl LateLintPass<'_> for ConfusingXorAndPow {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if_chain! {
+            if let ExprKind::Binary(op, left, right) = &expr.kind;
+            if op.node == BinOpKind::BitXor;
+            if let ExprKind::Lit(litr) = &right.kind;
+            if let ExprKind::Lit(litl) = &left.kind;
+            if let snip_left = snippet_opt(cx, litl.span).unwrap();
+            if let snip_right = snippet_opt(cx, litr.span).unwrap();
+            if get_numlit(litr, &snip_right)
+                        .unwrap()
+                        .is_decimal();
+            if get_numlit(litl, &snip_left)
+                        .unwrap()
+                        .is_decimal();
+                then {
+
+                            let left_val = unwrap_lit_to_dec(&left.kind).unwrap_or(0);
+                            let right_val = unwrap_lit_to_dec(&right.kind).unwrap_or(0);
+                            let suffix: &str = get_numlit( litr, &snip_right).unwrap().suffix.unwrap_or("");
+                            if left_val == 2 &&
+                             (right_val == 8 ||
+                            right_val == 16 ||
+                            right_val == 32 ||
+                            right_val == 64)
+                            {
+                                clippy_utils::diagnostics::span_lint_and_sugg(
+                                    cx,
+                                    CONFUSING_XOR_AND_POW,
+                                    expr.span,
+                                    "it appears that you are trying to get the maximum value of an integer, but '^' is not exponentiation operator",
+                                    "try with",
+                                    format!("u{right_val}::MAX"),
+                                    Applicability::MaybeIncorrect,
+                                );
+                            } else {
+                                // Even then, warn always.
+                                clippy_utils::diagnostics::span_lint_and_sugg(
+                                    cx,
+                                    CONFUSING_XOR_AND_POW,
+                                    expr.span,
+                                    "'^' is not the exponentiation operator",
+                                    "did you mean to write",
+                                    format!("{left_val}{suffix}.pow({right_val})"),
+                                    Applicability::MaybeIncorrect,
+                                );
+                            }
+
+                    }
+        }
+    }
+}
+
+// from ExprKind::Lit.node (Int) to u128
+fn unwrap_lit_to_dec(expr: &ExprKind<'_>) -> Option<u128> {
```

Comment:
since it's always called on expressions, you could change it to take `Expr` directly and check the kind in the body